packages/medusa/src/workflows/admin/craete-product/create-product.ts

```diff
@@ -0,0 +1,159 @@
+import { EntityManager } from "typeorm"
+import {
+  IInventoryService,
+  MedusaContainer,
+  ProductTypes,
+} from "@medusajs/types"
+import { ulid } from "ulid"
+import { MedusaError } from "@medusajs/utils"
+import {
+  DistributedTransaction,
+  TransactionHandlerType,
+  TransactionOrchestrator,
+  TransactionPayload,
+  TransactionState,
+  TransactionStepsDefinition,
+} from "../../../utils/transaction"
+import { CreateProductVariantInput } from "../../../types/product-variant"
+import {
+  attachInventoryItems,
+  createInventoryItems,
+  createProducts,
+  removeInventoryItems,
+  removeProducts,
+} from "../../functions"
+
+enum Actions {
+  createProduct = "createProduct",
+  createPrices = "createPrices",
+  attachToSalesChannel = "attachToSalesChannel",
+  createInventoryItems = "createInventoryItems",
+  attachInventoryItems = "attachInventoryItems",
+}
+
+const workflowSteps: TransactionStepsDefinition = {
+  next: {
+    action: Actions.createProduct,
+    saveResponse: true,
+    next: {
+      action: Actions.attachToSalesChannel,
+      saveResponse: true,
+      next: {
+        action: Actions.createPrices,
+        saveResponse: true,
+        next: {
+          action: Actions.createInventoryItems,
+          saveResponse: true,
+          next: {
+            action: Actions.attachInventoryItems,
+            noCompensation: true,
+          },
+        },
+      },
+    },
+  },
+}
+
+const createProductOrchestrator = new TransactionOrchestrator(
+  "create-product",
+  workflowSteps
+)
+
+type InjectedDependencies = {
+  manager: EntityManager
+  container: MedusaContainer
+  inventoryService?: IInventoryService
+}
+
+export async function createProductWorkflow(
+  dependencies: InjectedDependencies,
+  productId: string,
+  input: CreateProductVariantInput[]
+): Promise<DistributedTransaction> {
+  const { manager, container } = dependencies
+  async function transactionHandler(
+    actionId: string,
+    type: TransactionHandlerType,
+    payload: TransactionPayload
+  ) {
+    const command = {
+      [Actions.createProduct]: {
+        [TransactionHandlerType.INVOKE]: async (
+          data: ProductTypes.CreateProductDTO[]
+        ) => {
+          return await createProducts({
+            container,
+            data,
+          })
+        },
+        [TransactionHandlerType.COMPENSATE]: async (
+          data: any[],
+          { invoke }
+        ) => {
+          const createdProducts = invoke[Actions.createProduct]
+          return await removeProducts({ container, data: createdProducts })
+        },
+      },
```

Comment:
suggestion:

Because each of these functions aren't doing much else, maybe this?

```
[Actions.createProduct]: {
        [TransactionHandlerType.INVOKE]: createProductInvokeHandler,
        [TransactionHandlerType.COMPENSATE]: createProductCompensateHandler
      },
```

Reply:
also, I intentionally does not name with `Invoke` and `Compensate` because they can be used in different places. For example, if you have a delete product workflow, the remove product would be the invoke and the restoration would be the compensate. Therefore, the same function can be used either during an invokation or a compensation depending on the needs